singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -1417,25 +1457,26 @@ PORTABLE_INLINE_FUNCTION bool PTESolver(System &s) {
 
     // possibly scale the update to stay within reasonable bounds
     Real scale = s.ScaleDx();
-    // const Real scale_save = scale;
+    PORTABLE_REQUIRE(scale <= 1.0, "PTE Solver is attempting to increase the step size");
```

Comment:
when can this be reached?

Reply:
This is an attempt to catch future bugs like the one I found. We had _thought_ that the code as written would only ever produce a smaller step, but when you were on the other side of the minimum pressure density, it would actually attempt to _increase_ the step in order to move the PTE solver back to a higher density.

I read somewhere that the appropriate role of asserts is to ensure your code is acting as you expect it to act (i.e. in release it's likely a waste to actually check these things). I think this is a perfect instance where an assert is needed; we believed the code acted a certain way and were proven wrong.

There's a plausible argument that the asserts I have elsewhere should actually be hard errors, but I hesitate to impose that performance cost at the library level and instead I think the host codes can make that decision themselves before passing data to the PTE solver.

Reply:
Right if we believe this is impossible to reach, I think this should be an assert. My concern was that this might now happen naturally now that you've modified the line search to sometimes grow the line step size.

Reply:
This code is actually before the line search:
```c++
    // possibly scale the update to stay within reasonable bounds
    Real scale = s.ScaleDx();
    PORTABLE_REQUIRE(scale <= 1.0, "PTE Solver is attempting to increase the step size");

    // Line search
    Real gradfdx = -2.0 * scale * err;
    scale = 1.0; // New scale for line search
    Real err_old = err;
```

We reset the scale to 1.0 for the line search, and my read of the code is that the line search _should_ just limit the step.

My changes to the line search should be completely cosmetic and I don't believe the line search can increase the step beyond 1. I did add this comment and it might be less than clear. This code itself is not mine though:
```c++
      scale = 0.5;
      Real err_mid = s.TestUpdate(scale);
      if (err_mid < err && err_mid < err_old) {
        // try a larger step since a half step reduces the error
        scale = 0.75 + 0.5 * robust::ratio(err_mid - err, err - 2.0 * err_mid + err_old);
      }
```
I admit I didn't fully understand this code before, but after staring at it a bit more, I think I get it. I _think_ this code is basically assuming a roughly parabolic error and trying to guess where the minimum lies. This minimum is then used for the first iteration of the line search and _should_ be a pretty good guess.

But more specifically, I believe we can guarantee that `scale` lies somewhere between 0.25 and 0.75 when this code is executed. If `err_mid < err_old` and `err_mid < err` then the numerator is negative and `scale` can be at most 0.75 when `err_old` is very large compared to the difference between `err` and `err_mid`. On the other hand, if `err` is much larger than `err_old`, then the ratio goes to one and the scale will be at least 0.25.

If however the mid step is worse than the full step or the old error, then we just start at a full step (limited by `s.ScaleDx()`) and back up until we're decreasing the error by some appropriate amount.

I'll amend my comment to make this more clear as I assume this is why you thought my changes could increase the step somehow.

Reply:
@Yurlungur I'm going to resolve this. If there's something that needs to be resolved, let me know. But I think I'm fairly confident that this assert in question shouldn't be triggered

Reply:
👍 I'm happy with your explanation.